### PR TITLE
internal: fix a panic due to a racy Close routine

### LINKED ISSUE #150: panic: send on closed channel, from dride.io
Periodically, a customers ngrok 3.31 agent on Arm is crashing.

Link to screenshot, I cannot attach for some reason. 
[https://files.slack.com/files-pri/T051C43HT-F06C84KS9PB/image.png](https://files.slack.com/files-pri/T051C43HT-F06C84KS9PB/image.png
)

Here is the ZenDesk ticket: https://ngrok.zendesk.com/agent/tickets/16410

The customer also provided a agent log, but I think the screenshot might be enough to identify the problem.

```
INFO[12-28|17:58:50] no configuration paths supplied
DBUG[12-28|17:58:50] ngrok config file at legacy location does not exist legacy_path=/.config/ngrok/ngrok.yml
INFO[12-28|17:58:50] using configuration at default config path path=/.config/ngrok/ngrok.yml
INFO[12-28|17:58:50] open config file                         path=/.config/ngrok/ngrok.yml err=nil
t=2023-12-28T17:58:50+0000 lvl=dbug msg="ignoring manifest file with unhandled mimetype" name=assets/glyphicons-halflings-regular.ttf ext=.ttf
t=2023-12-28T17:58:50+0000 lvl=dbug msg="ignoring manifest file with unhandled mimetype" name=assets/glyphicons-halflings-regular.eot ext=.eot
t=2023-12-28T17:58:50+0000 lvl=warn msg="can't bind default web address, trying alternatives" obj=web addr=[127.0.0.1:4040](http://127.0.0.1:4040/)
t=2023-12-28T17:58:50+0000 lvl=dbug msg="starting component" obj=app name="Tunnel session"
t=2023-12-28T17:58:50+0000 lvl=dbug msg="starting component" obj=app name=web
t=2023-12-28T17:58:50+0000 lvl=info msg="starting web service" obj=web addr=[127.0.0.1:4041](http://127.0.0.1:4041/) allow_hosts=[]
t=2023-12-28T17:58:50+0000 lvl=dbug msg="starting component" obj=app name="signal handler"
t=2023-12-28T17:58:51+0000 lvl=dbug msg="connecting to" obj=tunnels.session addr=[connect.eu.ngrok-agent.com:443](http://connect.eu.ngrok-agent.com:443/)
t=2023-12-28T17:58:51+0000 lvl=dbug msg=dial obj=tunnels.session network=tcp address=[tunnel.eu.ngrok.com:443](http://tunnel.eu.ngrok.com:443/) timeout=10s
t=2023-12-28T17:58:51+0000 lvl=dbug msg="open stream" obj=tunnels.session obj=csess id=d0b9352af6f4 reqtype=0 err=nil
t=2023-12-28T17:58:51+0000 lvl=dbug msg="encode request" obj=tunnels.session obj=csess id=d0b9352af6f4 sid=3 req="&{Version:[2] ClientID: Extra:{OS:linux Arch:arm Authtoken:HIDDEN Version:3.3.1 Hostname: UserAgent:ngrok-agent-go/3.3.1 ({\"proxy_type\":\"none\",\"config_version\":\"2\"}) ngrok-go/1.1.0 Metadata: Cookie: HeartbeatInterval:10000000000 HeartbeatTolerance:15000000000 Fingerprint:<nil> UpdateUnsupportedError:0x280f368 StopUnsupportedError:0x280f2a8 RestartUnsupportedError:0x280f350 ProxyType: MutualTLS:false ServiceRun:false ConfigVersion: CustomInterface:false CustomCAs:false ClientType:ngrok-agent-go}}" err=nil
t=2023-12-28T17:58:51+0000 lvl=dbug msg="decoded response" obj=tunnels.session obj=csess id=d0b9352af6f4 sid=3 resp="&{Version:2 ClientID:3996dec297e9127d8d8ff0ce5ced7636 Error: Extra:{Version:prod Region:eu Cookie:BtnCf6OIZ5STrtvqBYhjZ6T9h585Rldq$zaNfuVpOBhTy8cIRCs9qBp/X29cX1Kx4/gaylYzitCELLiNqCHY8JatnCBfVBV/+29B8JJ+Zj7DxP72wuZfEbwmJcp5pSoLxQTlOQIDJdcNI0aw55Nv8MGReJKOq3IHO18IA6A8VCy+2T7WEm/ovFR0vn0C4qpCJFpUNxYJj7flXad6UtRmLzkU0EEXbHne+2xc8ZnmmVKYUfMIw0w== AccountName:Yossi Neiman SessionDuration:0 PlanName:Pro Banner:Build better APIs with ngrok. Early access: [ngrok.com/early-access}](http://ngrok.com/early-access%7D)}" err=nil
t=2023-12-28T17:58:51+0000 lvl=info msg="client session established" obj=tunnels.session obj=csess id=78d8bc87e742
t=2023-12-28T17:58:51+0000 lvl=info msg="tunnel session started" obj=tunnels.session
t=2023-12-28T17:58:51+0000 lvl=dbug msg="received extras" obj=tunnels.session region=eu version=prod
t=2023-12-28T17:58:51+0000 lvl=dbug msg="open stream" obj=tunnels.session obj=csess id=d0b9352af6f4 clientid=3996dec297e9127d8d8ff0ce5ced7636 reqtype=1 err=nil
t=2023-12-28T17:58:51+0000 lvl=dbug msg="encode request" obj=tunnels.session obj=csess id=d0b9352af6f4 clientid=3996dec297e9127d8d8ff0ce5ced7636 sid=7 req="&{ID: ClientID: Proto:https ForwardsTo:[http://localhost:80](http://localhost/) Opts:0x2d4c840 Extra:{Token: IPPolicyRef: Metadata:}}" err=nil
t=2023-12-28T17:58:51+0000 lvl=dbug msg="decoded response" obj=tunnels.session obj=csess id=d0b9352af6f4 clientid=3996dec297e9127d8d8ff0ce5ced7636 sid=7 resp="&{ClientID:06972402699776bfd0ebda17975a1883 URL:https://c7a22b6b26c6.ngrok.app/ Proto:https Opts:map[Auth: BasicAuth:<nil> CircuitBreaker:<nil> Compression:<nil> Domain: HostHeaderRewrite:false Hostname:[c7a22b6b26c6.ngrok.app](http://c7a22b6b26c6.ngrok.app/) IPRestriction:<nil> LocalURLScheme: MutualTLSCA:<nil> OAuth:<nil> OIDC:<nil> Policies:<nil> ProxyProto:0 RequestHeaders:<nil> ResponseHeaders:<nil> Subdomain: UpstreamCAPEM:<nil> UserAgentFilter:<nil> VerifyUpstreamTLS:false WebhookVerification:<nil> WebsocketTCPConverter:<nil>] Error: Extra:{Token:KXcIgeKA7BoTZ3z4574FivS4Lh18rJbs$wF7lKPVhfK2x/qX8BsnyK+i2l32CSb3JoR2tYLfbe1BxlyQZF17+iPvU1/A9L/Cgf7AiIbmyZnSSfyzX2P2Kjj0p1ihTHJHVXrEXE3mRBEj8SkoVdHrpVvKY3SZQV6YyLX4YRZsf3J/novSik2wbZCX/xR6zjzmvnfXcYMFHoRJBSKV6p9j700kdYVY5kLxzfI94U1y0CimUzcZIdx4pZejCYILLQ9IOopvrVl3vYfiu3tf2mU+BZuOX5rp6NmAfbY/0+ToKeIFXr3W2k6YeB9HKuI/vOtDTyj24WjwzCbfcSMcoYWiokB2faylMbSzuqQ6LGB6NK3gKnDfJLM3HjH4Cj26LJt0PSSONC4rzC6oECoQJkgH38sHdBh0SJxYlh0NxJnfBSZM9QJkpFzTnSfrezZsJKKGnAIhWANjOmfIf7jXd40k+Yfrm/qml+Y6MdcG4/Fsw3zK6p4X93rbVnydou+EZjDgavnsQFcHwq7lL/eX7w18lV36R8vtGDUCozBLqE50yuq1QE/G2Ag2N4j9csqeOcWht6P+b/5v/HdYwAIhFHqhb6x2B2EFP4erDY0I97PCkyVow4WisHuSo3mCw2PM+82Em3NN7nWnXAN5qIUc9w6bvurRbukf+A6fpdnnLYjqoO9NxKWryAsWkIp8AY5dU87lGiZvAG76VRDll4uUMZnRiaa/zwZWd8QN5qm4+mLCwe6oL1ekPB59n}}" err=nil
t=2023-12-28T17:58:51+0000 lvl=dbug msg="start tunnel listen" obj=tunnels.session name=command_line proto=https opts="&{Domain: Hostname: Subdomain: Auth: HostHeaderRewrite:false LocalURLScheme:http ProxyProto:0 Compression:<nil> CircuitBreaker:<nil> IPRestriction:<nil> BasicAuth:<nil> OAuth:<nil> OIDC:<nil> WebhookVerification:<nil> MutualTLSCA:<nil> RequestHeaders:<nil> ResponseHeaders:<nil> WebsocketTCPConverter:<nil>}" labels=map[] err=nil
t=2023-12-28T17:58:51+0000 lvl=info msg="started tunnel" obj=tunnels name=command_line addr=[http://localhost:80](http://localhost/) url=https://c7a22b6b26c6.ngrok.app/
udhcpd: sending OFFER of 192.168.10.182
udhcpd: sending OFFER of 192.168.10.182
t=2023-12-28T17:59:01+0000 lvl=dbug msg="heartbeat received" obj=tunnels.session obj=csess id=d0b9352af6f4 clientid=3996dec297e9127d8d8ff0ce5ced7636 latency_ms=100
udhcpd: sending OFFER of 192.168.10.182
t=2023-12-28T17:59:11+0000 lvl=dbug msg="heartbeat received" obj=tunnels.session obj=csess id=d0b9352af6f4 clientid=3996dec297e9127d8d8ff0ce5ced7636 latency_ms=114
t=2023-12-28T17:59:21+0000 lvl=dbug msg="heartbeat received" obj=tunnels.session obj=csess id=d0b9352af6f4 clientid=3996dec297e9127d8d8ff0ce5ced7636 latency_ms=140
udhcpd: sending OFFER of 192.168.10.182
^Ct=2023-12-28T17:59:28+0000 lvl=dbug msg="requested stop" obj=app
t=2023-12-28T17:59:28+0000 lvl=info msg="received stop request" obj=app stopReq="{err:<nil> restart:false}"
t=2023-12-28T17:59:28+0000 lvl=info msg="session closing" obj=tunnels.session err=nil
t=2023-12-28T17:59:28+0000 lvl=dbug msg="waiting for all components to stop" obj=app
t=2023-12-28T17:59:28+0000 lvl=dbug msg="waiting for components to stop" obj=app remaining=3
t=2023-12-28T17:59:28+0000 lvl=dbug msg="waiting for components to stop" obj=app remaining=2
t=2023-12-28T17:59:28+0000 lvl=dbug msg="waiting for components to stop" obj=app remaining=1
t=2023-12-28T17:59:28+0000 lvl=dbug msg="component stopped" obj=app name="Tunnel session" err=nil
t=2023-12-28T17:59:28+0000 lvl=dbug msg="component stopped" obj=app name=web err=nil
t=2023-12-28T17:59:28+0000 lvl=dbug msg="all components stopped" obj=app
```

### PR DESCRIPTION
**Pull request by jrobsonchase**
Resolves #150

This is a little awkward.

First off, we need to preserve the fact that the latency channel gets closed.
Downstream code relies on this property.

The big problem is that the process sending things on the latency channel is
completely out-of-band with regard to the Close method. Ideally, the Heartbeat
type would also take a "cleanup" closure that would only be called after all of
its goroutines have exited, but that's a muxado change that I don't really want
to make right now.

Instead, we just have to make `Close` and `onHeartbeat` race-free.

### CODE REVIEW

**Comment by jrobsonchase** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #151** <a href="https://app.graphite.dev/github/pr/ngrok/ngrok-go/151?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/ngrok/ngrok-go/151?utm_source=stack-comment).

File: internal/tunnel/client/raw_session.go [hunk 1]
```diff
@@ -46,11 +47,12 @@ type SessionHandler interface {
...
-	id               string            // session id for logging purposes
-	handler          SessionHandler    // callbacks to allow the application to handle requests from the server
-	latency          chan time.Duration
-	closeLatencyOnce sync.Once
+	mux     *muxado.Heartbeat // the muxado session we're multiplexing streams over
+	id      string            // session id for logging purposes
+	handler SessionHandler    // callbacks to allow the application to handle requests from the server
+	latency chan time.Duration
+	closed  atomic.Bool
+	cbGroup sync.WaitGroup
```
Comment (euank) [inline internal/tunnel/client/raw_session.go:55]: I feel like this juggling might be better modeled with the internal `reliable.Shutdown`, which we could pull out here.

Basically, I think if we do:

```go
func (s *rawSession) Close() {
  if s.shutdown.Shutdown() {
    close(s.latency)
  }
}

func (s *rawSession) onHeartbeat() {
  if timeout {
    s.shutdown.Shutdown()
    return
  }
  s.shutdown.Do(func() {
    // latency
  })
}
```

I think that ends up being identical to the code you have now, but with a slightly cleaner abstraction maybe?

Untested of course, so I'm not 100% confident. I do think having a unit test that reproduces the panic, and thus lets us be sure it's fixed, would be really helpful!

**Review by euank — COMMENTED** [review]
Looks good I think!

I have one suggestion for an alternate abstraction, but if you like this more, I'm happy to properly think through it and approve it too I think!

File: internal/tunnel/client/raw_session.go (hunk 1 above)
Comment (jrobsonchase) [reply internal/tunnel/client/raw_session.go]: I realized that a `RWLock` is a more direct way to represent this roundabout mutual exclusivity. Should be a lot cleaner now!

File: internal/tunnel/client/raw_session.go [hunk 2]
```diff
@@ -268,6 +278,15 @@ func (s *rawSession) rpc(reqtype proto.ReqType, req any, resp any) error {
 }
 
 func (s *rawSession) onHeartbeat(pingTime time.Duration, timeout bool) {
+	// make sure we don't send on a closed channel.
+	// Any number of `onHeartbeat` callbacks can be in flight at a given time,
+	// but only one Close.
+	s.closedLock.RLock()
```
Comment (euank) [inline internal/tunnel/client/raw_session.go:284]: This ends up with a deadlock in the `s.Close` before when it tries to take the `Lock`, right?

Manually scoping the lock by unlocking it before calling s.Close probably fixes that easily enough, and I think overall does make this simpler, good idea on modeling it this way
Still, unit tests feel very warranted here!

Comment (jrobsonchase) [reply internal/tunnel/client/raw_session.go:284]: Whoops, you are correct! Overlooked that `onHeartbeat` calls `Close` itself.

**Review by euank — APPROVED** [review]
LGTM, thanks for the unit test!

**Pull request merged**